### ISSUE TITLE
When your timezone is different from the user you're messaging, show their current time above/near the compose box

### ISSUE THREAD

**Issue by michaelhaxhiu** [Expensify/Expensify.cash#3819]
When your timezone is different from the user you're messaging, show **their** current time above the compose box (we already show their current time in their profile, this would just give you easier access to it).

**Coming from this slack thread**: https://expensify.slack.com/archives/C01GTK53T8Q/p1617771939406800

**Upwork job link** - https://www.upwork.com/jobs/~010fd4b2aae85f9fff

## Expected Result:
Show the time and timezone of the individual you're DMing when you're in a different timezone from them. 
Don't show this if it's a group DM (i.e. any DM with more than 1 user).

Let's use the format like so:
- PM: `It's 09:50 PM for Tom`
- AM: `It's 12:30 AM for Tom`

Here's a mockup of the proposed UI addition:
<img width=300 src=https://user-images.githubusercontent.com/13206896/123963841-d951f380-d980-11eb-9594-8826f0967460.png>

## Actual Result:
We don't show the time zone of the person you are messaging.

## Action Performed:

1. Open the Expensify Cash app or web app
2. Sign into your account
2. Go to any DM for a user that is not in your timezone
3. See that you have no indication that they are not in your timezone

## Platform:

All platforms

**Comment by MelvinBot** [comment]
Triggered auto assignment to @michaelhaxhiu (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by parasharrajat** [comment]
### Proposal
1. WE can get the user's Timezone details from ONYXKEYS.PERSONAL_DETAILS Onyx key. 
2. `moment().tz(userDetails.timezone.selected).format('LT')` will give us local time of that user.
3. We subscribe to the above key in ReportActionCompose.
4. We added a separate View to show the local time string above the composer chatbox. Above this code line https://github.com/Expensify/Expensify.cash/blob/c5552954fd4ebcef0874edc10feaa80456f1de35/src/pages/home/report/ReportActionCompose.js#L413
5. We will check the type of DM via length of participants which we are doing at https://github.com/Expensify/Expensify.cash/blob/c5552954fd4ebcef0874edc10feaa80456f1de35/src/pages/home/report/ReportActionCompose.js#L359.
6. We will pick the other user via participants.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @joelbettner (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by michaelhaxhiu** [comment]
Upwork Job is [posted](https://www.upwork.com/jobs/~010fd4b2aae85f9fff) and we should be fielding proposals shortly!

**Comment by Prince-Hamza** [comment]
Here is the code for converting into another time zone
function convertTZ(date, tzString) {
    return new Date((typeof date === "string" ? new 
    Date(date) : date).toLocaleString("en-US", 
    {timeZone: tzString}));   
}

const date = new Date()
convertTZ(date, "Asia/Jakarta")  //current date-time in jakarta.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @beamanator (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by michaelhaxhiu** [comment]
@Prince-Hamza did you want to propose the full solution to take this job? All good if not, just wanted to check!

**Comment by Prince-Hamza** [comment]
> @Prince-Hamza did you want to propose the full solution to take this job? All good if not, just wanted to check!

Yes i can give full solution. Can you explain what you want.

**Comment by parasharrajat** [comment]
Please find the Proposal https://github.com/Expensify/Expensify.cash/issues/3819#issuecomment-871426534. Thanks

**Comment by michaelhaxhiu** [comment]
@parasharrajat yep we have your proposal as the first one.

**Comment by Prince-Hamza** [comment]
this might be the full solution
`
componentDidMount = () => {
  Main()
}
getMyTimeZoneString = async () => {
  resp = await fetch('https://ipwhois.app/json/');
  user = await resp.json();
  return user.timezone // Asia/Karachi
};

function convertTZ (date, tzString) {
    return new Date((typeof date === "string" ? new
    Date(date) : date).toLocaleString("en-US",
    {timeZone: tzString}));
}

Main = async () => {
 var  myTimeZoneString = await getMyTimeZoneString() // get timezone from api or ONYX.PERSONAL
 var partnerTimeZone = new Date(2021, 1, 24, 10, 33, 30, 0); // get remote date here from database
 const isGroup = lodashGet(this.props.report, 'participants.length') == 2; 
 if (!isGroup) result = convertTZ(partnerTimeZone, myTimeZoneString)   // localize remote time if not group chat
this.setState({previewTime:result})
}
`

**Comment by MelvinBot** [comment]
Triggered auto assignment to @tgolen (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by Beamanator** [comment]
I'm taking this one @tgolen , I just added `Exported` since the job was already posted in Upwork

**Comment by Beamanator** [comment]
@parasharrajat Thanks for the proposal, I like your use of `moment-timezone` and subscribing to `ONYXKEYS.PERSONAL_DETAILS` in `ReportActionCompose` definitely makes sense. I have one comment: Can you clarify how steps 5 and 6 from your proposal will affect your implementation? I'm guessing you're planning to check `hasMultipleParticipants` so this field will only show up for 1:1 DMs, not group DMs - is that correct?

@Prince-Hamza Thanks for the proposal, but it looks like your proposal doesn't take into account any of the currently existing features of our application. For example, we can get a user's timezone using the key mentioned above. Also, since we use the `moment` library, it's usually easier to use that than to use the JavaScript `Date` object. I see you noticed we use `lodashGet` which is great, but it would also have been better if we could see which files you planned to add your implementation to.

**Comment by parasharrajat** [comment]
@Beamanator Yes. `hasMultipleParticipants` will be true for Group DMs only. For 1:1 it will contain other user's email.

**Comment by Beamanator** [comment]
@parasharrajat right, I'm trying to get you to confirm that you _will_ or _will not_ show timezones for Group DMs :)

**Comment by parasharrajat** [comment]
No Time will only be shown for 1:1 DM . i.e. when hasMultipleParticipanta is false.

**Comment by Beamanator** [comment]
Perfect, just making sure we're understanding each other. Thanks @parasharrajat - please submit a PR when you have a chance and apply for the job in Upwork!

**Comment by parasharrajat** [comment]
I have few questions:
1. Do we need to always show the user's time? What is the expected behaviour? I am not able to determine the exact behaviour of slack to correlate.
2. Now as time is a real-world thing so it should update at every moment to be correct. Do you want me to put a timer for this time if we are always going to show this?

**Comment by Beamanator** [comment]
> 1. Do we need to always show the user's time? What is the expected behaviour? I am not able to determine the exact behaviour of slack to correlate.

It seems like Slack only shows the other user's local time if their time is out of normal business hours, according to [the linked Slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1617805410415600?thread_ts=1617771939.406800&cid=C01GTK53T8Q). Can you ask there for exact implementation details? (ex: show the user's local time if it's not 9am - 5pm there, or show their timezone always, if different from your own)

> 2. Now as time is a real-world thing so it should update at every moment to be correct. Do you want me to put a timer for this time if we are always going to show this?

Good question, yes I do believe this needs to be done so the user's local time stays up to date

**Comment by michaelhaxhiu** [comment]
Instead of `9:50pm Local time`, let's use the format `It's 9:50 PM for Tom` as discussed [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1625559881478500?thread_ts=1617771939.406800&cid=C01GTK53T8Q) in slack.

**Comment by michaelhaxhiu** [comment]
Looks like we are waiting on Expensify/Expensify.cash#3958. I'll test this when it's live.

Payment will be due 7 days after deployment.

**Comment by parasharrajat** [comment]
@michaelhaxhiu This is already deployed and you must be seeing the message over composer `Its xx:xx For User`

**Comment by michaelhaxhiu** [comment]
Ah sorry to clarify, I'm waiting to test this on `production` on the desktop app specifically. I see it on TF on mobile.

**Comment by rushatgabhane** [comment]
`It's 09:50 AM for Tom` shows even if my time zone is same as Tom's.

Timezone should only be shown if it's different.

**Comment by Beamanator** [comment]
Hmm can you try hard-refreshing both chats? I am not experiencing this in staging

**Comment by rushatgabhane** [comment]
> @rushatgabhane Hmm can you try hard-refreshing both chats? I am not experiencing this in staging

Hard-refresh fixed it, my bad 😅

**Comment by rdjuric** [comment]
@parasharrajat I saw that https://github.com/Expensify/Expensify.cash/pull/3942 was merged, but the text still isn't wrapping/shrinking properly when the name is bigger than the container. The date is also wrapping when it shouldn't now. 

<img width="406" alt="Screen Shot 2021-07-14 at 9 28 33 PM" src="https://user-images.githubusercontent.com/53711423/125767848-fa3c1848-c207-4de6-ab8c-c249c2c527cf.png">

**Comment by parasharrajat** [comment]
Oh thanks for bringing this to attention. Its is a regression after we made name and time to bold. No worries I will pick this up with ongoing change requests.

**Comment by michaelhaxhiu** [comment]
>@parasharrajat I saw that #3942 was merged, but the text still isn't wrapping/shrinking properly when the name is bigger than the container. The date is also wrapping when it shouldn't now.

I don't see a GH for yet -- did we make one yet?

**Comment by parasharrajat** [comment]
So this is a regression from Rory's changes where he added boldness to Name and time. But I will tackle it as I have another change regarding the same. 

@michaelhaxhiu May be use can adjust this https://github.com/Expensify/Expensify.cash/issues/3987 to reflect both changes.

**Comment by michaelhaxhiu** [comment]
@roryabraham what do you think? Make a new issue to fix the bold or add it to https://github.com/Expensify/Expensify.cash/issues/3987

**Comment by roryabraham** [comment]
I actually think that there was [consensus in Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1626211434198100) to remove all boldness from the timestamp and typing indicator. But yeah, can we just add it to https://github.com/Expensify/Expensify.cash/issues/3987?

**Comment by parasharrajat** [comment]
@michaelhaxhiu Any update on the Upwork job for this? Also could we please follow this https://github.com/Expensify/Expensify.cash/issues/3819#issuecomment-881052637 ? I will submit the fix asap.

**Comment by michaelhaxhiu** [comment]
@Expensify/contributor-management I'm offline right now and need a buddy to help unblock @parasharrajat . Looks like @roryabraham is OOO right now!

**Comment by mallenexpensify** [comment]
Hired then paid in Upwork, should be all good here

**Comment by michaelhaxhiu** [comment]
I updated https://github.com/Expensify/App/issues/3987 to include the discussed changes.
